Enable cross-compilation between both Darwin platforms

reckenrode opened the pull request:
## Description of changes

While cross-compilation from x86_64-darwin to aarch64-darwin already works, going the other way runs into issues with the source-based SDK. This PR resolves those issues and enables cross-compilation to x86_64-darwin. It should now be possible to build any of the following scenarios from a Darwin host:

* aarch64-darwin native (dynamic and static)
* x86-darwin native (dynamic and static)
* aarch64-darwin to x86_64-darwin (dynamic and static)
* x86_64-darwin to aarch64-darwin (dynamic and static)

I have tried to keep the changes required small and limited. While working on this PR, I found a few derivations that did not build correctly when cross-compiling. Fixing those was outside the scope of this PR unless they were required by the bootstrap or are part of the cross-stdenv.

Known breakage: brotli (references a dylib from the build environment), Wine (has a `buildInput` in one of its phase scripts), DarwinTools (needs modifications for cross-compilation). MoltenVK is probably also broken.

Regarding Linux: I briefly looked at whether this would work on Linux. There’s work needing to be done to make certain required tools like bootstrap_cmds work on Linux. Expanding cross-compilation to Linux is outside the scope of this PR.

**Note:** My tests were done using my clang 16 branch. I have successfully evaluated the clang 11 branch used for this PR, but it will take overnight to build. I wanted to open this PR to get it out there for review, so I am opening it as a draft. If all goes well, I’ll flip this to ready tomorrow (Friday, 2023-09-22).

### Resolved Issues

This PR should close the following issues:

* Fixes #126829
* Fixes #180771
* Closes #214611
* Fixes #273442

### Prerequisite PRs

### Related PRs

- [ ] https://github.com/NixOS/nixpkgs/pull/315180

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

reckenrode committed:
darwin.xnu: fix cross-compilation and cleanup

* Break an infinite recursion by determining the target prefix from
  the platform config. Referencing `stdenv.cc.bintools.targetPrefix`
  causes an infinite recursion because xnu is a dependency of Libsystem;
* Add the host compiler as a dependency for the headers-only build,
  which needs to build tools it will use at build time; and
* Add `migcc`, which provides a target compiler for use with `mig`,
  which needs to invoke clang for the target system during the build.

reckenrode committed:
darwin.Libsystem: fix cross-builds and cleanup

* Always use `stdenvNoCC` to build Libsystem. This is needed to prevent
  an infinite recursion when cross-building Libsystem;
* Drop ncurses from Libsystem. While upstream does ship ncurses headers,
  everything in nixpkgs should be using ncurses from nixpkgs and not
  using upstream’s headers. This fixes an infinite recursion because
  ncurses still needs to be build to produce an ncurses^dev output;
* Allow the stubs to be conditionally included in Libsystem. This is
  needed to build a bootstrap stdenv for building libresolv and Csu;
* Symlink libresolv instead of copying it into Libsystem’s output. It’s
  not clear why this is necessary anymore, and using a stdenv with
  `install_name_tool` causes an infinite recursion when cross-building.

reckenrode committed:
darwin.LibsystemCross: define on x86_64-darwin

reckenrode committed:
darwin.Csu: fix cross-compilation

Csu only needs to build stub objects, so it only needs to use the
bootstrap stdenv. This breaks an infinite recursion when cross-compiling
to x86_64-darwin.

reckenrode committed:
darwin.configd: fix cross-compilation

configd does not require a full Libsystem to build. Break an infinite
recursion by using the bootstrap stdenv.

reckenrode committed:
darwin.libresolv: fix cross-compilation

Break an infinite recursion by using the bootstrap stdenv, which
disables libresolv support in Libsystem.

reckenrode committed:
swift-corelibs: enable static builds of CF

reckenrode committed:
darwin.objc4: use stdenvNoCF to support Darwin cross

reckenrode committed:
swift-corelibs: use stdenvNoCF to support cross-compilation

reckenrode committed:
stdenv: use darwin.CF with Darwin cross-compilation

With the infinite recursion broken when including darwin.CF in the
stdenv, both regular cross and static builds can use the same
`extraBuildInputs` override.

reckenrode committed:
darwin.Libsystem: add feature flag compatibility with 10.12 SDK

reckenrode committed:
darwin.stdenvBootstrap: add for building Libsystem

Libsystem depends (directly and indirectly) on a few packages that
require clang to build. This stdenv uses a Libsystem that omits these
packages, so the one used by the rest of nixpkgs can include them.

reckenrode commented:
The clang 11 builds completed successfully after I added the fix for compiler-rt. Setting this to ready to review.

toonn reviewed (commented):
Mostly have minor remarks and questions due to unfamiliarity.

Awesome work once again! :heart:

reckenrode commented on pkgs/development/compilers/llvm/11/default.nix:
```diff
@@ -248,9 +249,12 @@ let
+      # Build compiler-rt without CF to prevent an infinite recursion when cross-compiling to Darwin.
+      stdenv = if stdenv.hostPlatform.isDarwin && (stdenv.buildPlatform != stdenv.hostPlatform)
+        then overrideCC darwin.stdenvNoCF buildLlvmTools.clangNoCompilerRtWithLibc
+        else if stdenv.hostPlatform.useLLVM or false || stdenv.isDarwin || (stdenv.hostPlatform.isRiscV && stdenv.hostPlatform.is32bit)
```
It still does. Note that line 255 includes `stdenv.isDarwin`. This is mirroring the logic from LLVM 16.

reckenrode replied:
The reason for having these two cases is because building without CF prevents building the sanitizers (because ASAN requires CoreFoundation). Ideally, the native builds would have sanitizers. This actually fixes aarch64-darwin to build sanitizers since it’s not building them currently.

toonn replied:
So it is intentional for native aarch64-darwin to not match the overrideCC case then. Good.

reckenrode commented on pkgs/development/compilers/llvm/11/default.nix:
```diff
@@ -271,23 +275,29 @@ let
-               else stdenv;
+      stdenv = if stdenv.hostPlatform.isDarwin && (stdenv.buildPlatform != stdenv.hostPlatform)
+        then overrideCC darwin.stdenvNoCF buildLlvmTools.clangNoLibcxx
+        else if stdenv.hostPlatform.useLLVM or false || stdenv.isDarwin
```
I made the change to match how the newer LLVM derivations are written. It may not strictly need it, but it’s easier when both platforms follow the same code paths.

toonn replied:
Minor issue anyway, fine.

reckenrode commented on pkgs/development/compilers/llvm/11/default.nix:
```diff
@@ -271,23 +275,29 @@ let
-      stdenv = if (stdenv.hostPlatform.useLLVM or false) || (stdenv.hostPlatform.isDarwin && stdenv.hostPlatform.isAarch64)
-               then overrideCC stdenv buildLlvmTools.clangNoLibcxx
-               else stdenv;
+      stdenv = if stdenv.hostPlatform.isDarwin && (stdenv.buildPlatform != stdenv.hostPlatform)
```
Add an `isDarwinCross` like the compiler-rt derivations have? If so, I can also mirror the logic to the LLVM 16 changes.

toonn replied:
Something like that, yes. I don't think there's a convention to make it clear it's build and host differing, while target doesn't matter.

Mindavi replied:
The convention is to keep it `(stdenv.buildPlatform != stdenv.hostPlatform)` to make sure it's clear what _kind_ of cross this is.

reckenrode replied:
In that case, should the compiler-rt changes also follow this pattern?

Mindavi replied:
I think it's ok as long as it's in the same file, actually.

reckenrode replied:
Ah, okay. I kept the changes localized to the relevant files. I’ll push the updated PR later tonight. It also includes a fix for Csu, which broke the Wine build.

reckenrode commented on pkgs/development/compilers/llvm/16/compiler-rt/default.nix:
```diff
@@ -49,15 +49,15 @@ stdenv.mkDerivation {
   ] ++ lib.optionals (haveLibc && stdenv.hostPlatform.libc == "glibc") [
     "-DSANITIZER_COMMON_CFLAGS=-I${libxcrypt}/include"
-  ] ++ lib.optionals (useLLVM || bareMetal || isMusl || isDarwinStatic) [
+  ] ++ lib.optionals (useLLVM || bareMetal || isMusl || isDarwinCross) [
```
Native static is a form of cross-compilation, so it gets included in this check.

toonn replied:
Ah, my bad. Makes sense, I guess.

reckenrode commented on pkgs/development/compilers/llvm/16/default.nix:
```diff
@@ -317,9 +317,12 @@ in let
-               then overrideCC stdenv buildLlvmTools.clangNoCompilerRtWithLibc
-               else stdenv;
+      # Build compiler-rt without CF to prevent an infinite recursion when cross-compiling to Darwin.
+      stdenv = if stdenv.hostPlatform.isDarwin && (stdenv.buildPlatform != stdenv.hostPlatform)
```
See above. If the idea is to define an `isDarwinCross`, I’ll do it the same way for both.

toonn replied:
The original idea here was just not having to index stdenv over and over but shorter is probably even better.

reckenrode commented on pkgs/development/compilers/llvm/16/default.nix:
```diff
@@ -317,9 +317,12 @@ in let
+      stdenv = if stdenv.hostPlatform.isDarwin && (stdenv.buildPlatform != stdenv.hostPlatform)
+        then overrideCC darwin.stdenvNoCF buildLlvmTools.clangNoCompilerRtWithLibc
+        else if stdenv.hostPlatform.useLLVM or false || stdenv.isDarwin
+        then overrideCC stdenv buildLlvmTools.clangNoCompilerRtWithLibc
```
Yes. Otherwise, the compiler-rt will try to build with a stdenv that requires compiler-rt.

reckenrode commented on pkgs/os-specific/darwin/apple-sdk-11.0/libSystem.nix:
```diff
@@ -1,4 +1,4 @@
-{ stdenvNoCC, buildPackages, MacOSX-SDK }:
+{ stdenvNoCC, buildPackages, MacOSX-SDK, withLibresolv ? null }:
```
It simplifies the definition of the bootstrap stdenv. It needs to override Libsystem not to include Csu and libresolv. It’s easier and cleaner if I can do that unconditionally (especially since Libsystem is versioned inconsistently depending on whether it comes from the SDK or is built from a source release).

toonn replied:
Maybe we should add a comment explaining the reasoning?
```suggestion
{ stdenvNoCC, buildPackages, MacOSX-SDK
  # Enables unconditional libSystem override
, withLibresolv ? null
}:
```

reckenrode commented on pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix:
```diff
@@ -161,17 +187,7 @@ appleDerivation' stdenv {
     done
   '' + lib.optionalString withLibresolv ''
-
-    # This probably doesn't belong here, but we want to stay similar to glibc, which includes resolv internally...
```
I can restore the comment.

reckenrode commented on pkgs/os-specific/darwin/apple-source-releases/configd/default.nix:
```diff
@@ -1,11 +1,23 @@
 
   nativeBuildInputs = lib.optionals (!headersOnly) [ bootstrap_cmds ];
-  buildInputs = lib.optionals (!headersOnly) [ launchd ppp xpc IOKit eap8021x ];
+  buildInputs = lib.optionals (!headersOnly) [ CF launchd ppp xpc IOKit eap8021x ];
```
I can add a comment explaining. The bootstrap stdenv doesn’t include CF, libresolv, or Csu. configd needs to use it because it’s a dependency of libresolv, but it also does need CF, so it adds it back as a `buildInput`.

reckenrode commented on pkgs/os-specific/darwin/apple-source-releases/libresolv/default.nix:
```diff
@@ -1,8 +1,17 @@
+, Libsystem
+, stdenvNoCC
+, stdenvBootstrap
+, buildPackages
+, overrideCC
```
Probably extra junk from when I was trying different things to break the infinite recursion. I can remove these.

toonn replied:
```suggestion
```

reckenrode commented on pkgs/os-specific/darwin/apple-source-releases/xnu/default.nix:
```diff
@@ -1,16 +1,40 @@
+  let
+    inherit (stdenv) hostPlatform targetPlatform;
+    arch = stdenv.hostPlatform.darwinArch;
+    targetPrefix = lib.optionalString (targetPlatform != hostPlatform) "${targetPlatform.config}-";
```
It should, but we can’t touch `cc` without causing an infinite recursion.

reckenrode commented on pkgs/os-specific/darwin/apple-source-releases/xnu/default.nix:
```diff
@@ -56,27 +80,40 @@ appleDerivation' (if headersOnly then stdenvNoCC else stdenv) (
-
-  env.NIX_CFLAGS_COMPILE = "-Wno-error";
+  env = {
+    inherit headersOnly;
```
That’s leftover from debugging the environment. I’ll remove it.

toonn replied:
```suggestion
```

reckenrode commented on pkgs/os-specific/darwin/swift-corelibs/corefoundation.nix:
```diff
@@ -1,4 +1,18 @@
-{ lib, stdenv, fetchFromGitHub, fetchurl, makeSetupHook, cmake, ninja, pkg-config, launchd, libdispatch, python3, libxml2, objc4, icu }:
```
Correct. It should have been removed when 8bee297 made the change to include the hook directly in CF’s output.

reckenrode commented on pkgs/stdenv/adapters.nix:
```diff
@@ -93,8 +93,6 @@ rec {
   # Best effort static binaries. Will still be linked to libSystem,
   # but more portable than Nix store binaries.
   makeStaticDarwin = stdenv: stdenv.override (old: {
-    # extraBuildInputs are dropped in cross.nix, but darwin still needs them
```
This is actually doing two things. First, this should be `buildPackages.targetPackages.darwin.CF`. Otherwise, static x86_64-darwin gets a dynamic CF instead of a static CF. Second, since that mirrors what `cross.nix` should have, I thought it would be cleaner to delegate to the override there instead of special casing for static.

toonn replied:
Are build- and targetPackages analogous to the build, host, target triumvirate? And if so, why wouldn't hostPackages suffice?

reckenrode replied:
There’s no `hostPackages`, but I can use `pkgsHostTarget` (assuming the current pattern isn’t somehow avoiding an infinite recursion).

reckenrode replied:
Wait, never mind. The way the cross stdenv is written doesn’t expose `pkgsHostTarget`. It only provides `buildPackages`.

reckenrode commented on pkgs/stdenv/darwin/default.nix:
```diff
@@ -320,15 +325,24 @@ in
               ln -s libSystem.tbd $out/lib/lib$name.tbd
             done
 
-            ln -s ${bootstrapTools}/lib/*.o $out/lib
```
They’re actually linked later, so it’s actually doing extra work. What I’m trying to do is align the stdenv bootstrap stage 0 with how the packages are actually organized/built.

toonn commented on pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix:
```diff
@@ -161,17 +187,7 @@ appleDerivation' stdenv {
     done
   '' + lib.optionalString withLibresolv ''
-
-    # This probably doesn't belong here, but we want to stay similar to glibc, which includes resolv internally...
```
```suggestion
    # This probably doesn't belong here, but we want to stay similar to glibc, which includes resolv internally...
```

reckenrode commented:
@toonn I believe I’ve addressed all the feedback. I tested static and dynamic cross builds to both x86_64-darwin and aarch64-darwin. I also confirmed Wine builds with the clang 16 stdenv (it does not build with the clang 11 one currently).

Et7f3 commented:
darwin-cross is at commit a8d8c7640fa324fca6c24ebd0ea7660957a2e556

`nix build 'git+file:.?ref=darwin-cross#pkgsCross.aarch64-darwin.darwin.binutils-unwrapped'` from x86_64-darwin with sandbox enabled wasn't able to build. At least two derivation failed.

`nix log 'git+file:.?ref=darwin-cross#pkgsCross.aarch64-darwin.python3' > pkgsCross.aarch64-darwin.python3.log`
`nix log 'git+file:.?ref=darwin-cross#pkgsCross.aarch64-darwin.llvm' > pkgsCross.aarch64-darwin.llvm.log`

[pkgsCross.aarch64-darwin.llvm.log](https://github.com/NixOS/nixpkgs/files/12841452/pkgsCross.aarch64-darwin.llvm.log)
[pkgsCross.aarch64-darwin.python3.log](https://github.com/NixOS/nixpkgs/files/12841453/pkgsCross.aarch64-darwin.python3.log)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-macos-monthly/12330/42

reckenrode commented:
> darwin-cross is at commit [a8d8c76](https://github.com/NixOS/nixpkgs/commit/a8d8c7640fa324fca6c24ebd0ea7660957a2e556)
> 
> `nix build 'git+file:.?ref=darwin-cross#pkgsCross.aarch64-darwin.darwin.binutils-unwrapped'` from x86_64-darwin with sandbox enabled wasn't able to build. At least two derivation failed.
> 
> `nix log 'git+file:.?ref=darwin-cross#pkgsCross.aarch64-darwin.python3' > pkgsCross.aarch64-darwin.python3.log` `nix log 'git+file:.?ref=darwin-cross#pkgsCross.aarch64-darwin.llvm' > pkgsCross.aarch64-darwin.llvm.log`
> 
> [pkgsCross.aarch64-darwin.llvm.log](https://github.com/NixOS/nixpkgs/files/12841452/pkgsCross.aarch64-darwin.llvm.log) [pkgsCross.aarch64-darwin.python3.log](https://github.com/NixOS/nixpkgs/files/12841453/pkgsCross.aarch64-darwin.python3.log) expected behavior [python/cpython#90905](https://github.com/python/cpython/issues/90905)
> 
> `pkgsCross.aarch64-darwin.libcxx` is fixed with this PR 👍🏻 The llvm error seems to be solvable. What solved is this change https://github.com/NixOS/nixpkgs/pull/256590/files#diff-dd802e439c9a73b6ade217547e77e33986a0dcf332d0c61337aead7e6ba1f1b2L89-R89

I finally got a chance to dig into the LLVM issue. I’ve been testing on my clang 16 branch, so I had to rebuild this branch for clang 11 to test and confirm, and that took a little bit. This is the issue.

https://github.com/NixOS/nixpkgs/blob/a68655c5eb93bc593f23a16b64a63edf9d2d9d21/pkgs/development/compilers/llvm/11/llvm/default.nix#L267-L271

LLVM needs a native compiler when doing a cross build. That is set up correctly, but the native compiler also needs certain dependencies propagated to it (such as the path to `libc++abi.dylib`), which does not happen. I have a patch for that, but since cross-compiling LLVM itself is not strictly necessary, it’s out of scope for this PR.

However, I do have a patch, but I want to make sure it applies and works across all of the LLVM versions. I’ll submit the PR (probably tomorrow morning EDT) for it once those finish building (for both aarch64-darwin and x86_64-darwin cross).

reckenrode commented:
I opened #260543 for the cross-compilation fix.

toonn commented on pkgs/os-specific/darwin/apple-source-releases/Csu/default.nix:
```diff
@@ -1,7 +1,10 @@
-
-appleDerivation' stdenv {
+{ lib
+, Libsystem
```
This isn't used and I don't think it serves the same unconditional override purpose?

toonn commented on pkgs/os-specific/darwin/apple-source-releases/libresolv/default.nix:
```diff
@@ -1,8 +1,15 @@
-{ lib, appleDerivation', stdenv, stdenvNoCC, Libinfo, configdHeaders, mDNSResponder
+{ lib
+, Libsystem
```
I must be missing the reason for these Libsystem arguments?

toonn reviewed (commented):
Noticed a couple more extra arguments, might be missing the reason.

toonn commented:
Let's also rebase this for the LLVM 16 stdenv.

reckenrode commented:
Sure. I’ll take a look at the feedback and push an update after this staging stuff settles down.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-macos-monthly/12330/44

avdv commented:
Just to note, I was trying if this PR would fix an issue with `pkgsStatic.coreutils` (see https://github.com/NixOS/nixpkgs/issues/270375), but it seems that it changed `pkgsStatic` on Darwin back to being equivalent to the normal packages set:
```console
$ nix-shell --argstr localSystem x86_64-darwin -I nixpkgs=./. --pure -p which pkgsStatic.coreutils --run 'which coreutils'
/nix/store/fbzhv146kg8b8mp90q8ncrqj7l2fj4wy-coreutils-9.3/bin/coreutils
$ nix-shell --argstr localSystem x86_64-darwin -I nixpkgs=./. --pure -p which coreutils --run 'which coreutils'
/nix/store/fbzhv146kg8b8mp90q8ncrqj7l2fj4wy-coreutils-9.3/bin/coreutils
```

reckenrode commented:
Thanks for the test case. The intent certainly is to get a (mostly) static version. I’ll be picking this PR back up over the next month or so. There were some changes made to Darwin that should allow it to be simplified a bit.

reckenrode committed:
darwin.apple_sdk_10_12: use stdenvNoCC instead of stdenv

The SDK frameworks don’t need a compiler in the general case, so don’t
use one unnecessarily. This avoids an infinite recursion when
cross-compiling to x86_64-darwin.

reckenrode committed:
darwin.apple_sdk.frameworks.CoreFoundation: drop hook

As of https://github.com/NixOS/nixpkgs/pull/265102, x86_64-darwin no
longer uses the open-source CF release. Since there is no longer a need
to switch between implementations, and the hook causes problems with
cross-compilation (see https://github.com/NixOS/nixpkgs/issues/278348),
drop the hook and make both darwin.CF an alias for
darwin.apple_sdk.frameworks.CoreFoundation.

reckenrode committed:
bacula: fix build on Darwin

After the CoreFoundation hook was dropped, bacula requires the following
changes to build on Darwin:

* Ensure bacula links CoreFoundation. Override incorrectly failing
  `configure` tests and add gettext as a build input. Even if those
  tests pass, bacula only links CoreFoundation when it finds libintl.
* Add Kerberos framework. Required to build on x86_64-darwin.

reckenrode committed:
librsvg: use libobjc on both Darwin architectures

reckenrode commented:
Thanks for the feedback. I’m currently waiting on https://github.com/NixOS/nixpkgs/pull/292043 before rebasing to take its changes into account.

reckenrode committed:
meson: fix extraframework dependencies on case-sensitive APFS

reckenrode committed:
libtapi: 1100.0.11 -> 1500.0.12.3

reckenrode committed:
cctools: 973.0.1 -> 1010.6

reckenrode committed:
darwin.stdenv: build llvmPackages.libunwind along with libc++ and libc++abi

reckenrode committed:
ld64: init at 951.9

reckenrode committed:
darwin.libdispatch: 442.1.4 -> 703.50.37

reckenrode committed:
ld-wrapper: use a temporary file for reponse file

This changes ld-wrapper to use a temporary file for the response file
passed to ld instead of using process substitution.

ld64 does not handle long command-lines when reading from the response
file, which defeats the point of using a response file to handle long
command-lines. cctools-port was patched to work around this, but nixpkgs
is now using Apple’s source release directly instead of the port.

Since it’s preferable not to patch Apple’s release heavily (to reduce
the difficulty of updating to new versions and to match upstream’s
behavior), use the approach that was adopted in cc-wrapper to work
around issues with response files in newer versions of clang.

Related PRs (cctools-port):
- https://github.com/NixOS/nixpkgs/pull/213831
- https://github.com/tpoechtrager/cctools-port/pull/132

Related PRs (cc-wrapper):
- https://github.com/NixOS/nixpkgs/pull/245282
- https://github.com/NixOS/nixpkgs/pull/258608

reckenrode committed:
darwin.configd: remove dependency on CF for private headers

reckenrode committed:
darwin.Libsystem: use lib.getDev instead of assuming the output exists

reckenrode committed:
binutils-wrapper: drop postLinkSignHook

When the linker signs a Mach-O binary, it sets a flag in the signature’s
code directory indicating that the signature was generated by a linker.
Tools such as `strip` and `install_name_tool` read this flag and will
update ad hoc signatures after they perform their modifications.

The updated l64 supports signing binaries automatically. Both the
updated cctools and LLVM will check for the linker-signed flag and
resign binaries they modify automatically when it’s present. Given that,
use of postLinkSignHook is unnecessary and potentially harmful.

In particular, if the hook is used and an unwrapped `strip` or
`install_name_tool` is on the user’s path, they will not automatically
update an ad hoc signature. Instead, they will issue a warning and
create a binary with a broken signature.

It is more robust to let the tools handled this since the only time a
signature would not be linker-signed is when the user is manually
invoking `codesign` (or another tool such as `sigtool` or `rcodesign`),
which by nature of the invocation updates the signature to a valid  one.

Since `strip` no longer needs to be wrapped for code-signing,
binutils-wrapper now uses the GNU strip wrapper on Darwin.

Fixes https://github.com/NixOS/nixpkgs/issues/208951.

reckenrode committed:
darwin.stdenv: update for updated cctools and ld64 and cleanup

- Use postLinkSignHook and wrap `strip` and `install_name_tool` until
  the updated cctools and ld64 can be built, which handles ad hoc
  signatures automatically as long as they are linker-signed;
- Drop postLinkSignHook from binutils once it is no longer needed;
- Move the minimal Python 3 with scproxy build to stage 1;
- Combine multi-part stage 2 into a single stage;
- Always use response files now that the bootstrap tools linker is a new
  enough version to support them;
- Use superDarwin.binutils-unwrapped even in stage0 because it just
  symlinks things into place. This simplifies to bootstrap;
- Drop use of cctools-llvm because the logic was moved to
  darwin.binutils-unwrapped, and cctools-llvm is now an alias for that;
- Prevent curl and GNU binutils from being used in the bootstrap; and
- Build libunwind along with libc++ and propagate it to the final
  overlay. It will eventually become the default libunwind on Darwin.

reckenrode committed:
fixDarwinDylibNames: use darwin.binutils not GNU binutils

reckenrode committed:
darwin.Libsytem: use top-level cctools

reckenrode committed:
darwin.binutilsDualAs: use top-level cctools

reckenrode committed:
libredirect: use llvmPackages unconditionally

Darwin and its bootstrap tools both use LLVM 16, which is sufficient for
arm64e support in LLVM. Thsi change removes an unnecessary LLVM 13 build
from the stdenv bootstrap.

reckenrode committed:
libunwind: use llvmPackages.libunwind on Darwin

reckenrode committed:
darwin.libunwind: remove and add to darwin-alaises

reckenrode committed:
dotnet: get dwarfdump from ld64

reckenrode committed:
xcbuild: use top-level cctools

reckenrode committed:
darwin.sigtool: use top-level cctools

reckenrode committed:
chez-racket: use top-level cctools

reckenrode committed:
python3Packages.jaxlib-build: use top-level cctools

reckenrode committed:
python3Packages.miniupnpc: use top-level cctools

reckenrode committed:
python3Packages.tensorflow: use top-level cctools

reckenrode committed:
python3Packages.tensorflow-probability: use top-level cctools

reckenrode committed:
clightning: use top-level cctools

reckenrode committed:
macvim: use top-level libunwind

reckenrode committed:
thelounge: use top-level cctools

reckenrode committed:
n8n: use top-level cctools

reckenrode committed:
build-npm-package: use top-level cctools

reckenrode committed:
bitwarden-cli: use top-level cctools

reckenrode committed:
frankenphp: use top-level cctools

reckenrode committed:
libplacebo_5: use top-level cctools

reckenrode committed:
live555: use top-level cctools

reckenrode committed:
protoc-gen-js: use top-level cctools

reckenrode committed:
iosevka: use top-level cctools

reckenrode committed:
dart: use top-level cctools

reckenrode committed:
elmPackages.nodejs: use top-level cctools

reckenrode committed:
gcc: use top-level cctools

reckenrode committed:
gcc: update reference to replacement package

reckenrode committed:
ponyc: use top-level cctools

reckenrode committed:
swift: use top-level cctools

reckenrode committed:
cyclone: use top-level cctools

reckenrode committed:
python: use top-level libunwind

reckenrode committed:
freeimage: use top-level cctools

reckenrode committed:
gstreamer: enable libunwind on Darwin

reckenrode committed:
libmcrypt: use top-level cctools

reckenrode committed:
libplacebo: enable libunwind on Darwin

reckenrode committed:
nss: use top-level cctools

reckenrode committed:
qt5: use top-level cctools and libunwind

reckenrode committed:
qt6: use top-level cctools and libunwind

reckenrode committed:
mkl: use top-level cctools

reckenrode committed:
petsc: use top-level cctools

reckenrode committed:
nodePackages: use top-level cctools

reckenrode committed:
rPackages.pbdZMQ: use top-level cctools

reckenrode committed:
bazel_{5,6,7}: use top-level cctools and ld64

reckenrode committed:
gn: use top-level cctools

reckenrode committed:
nodehun: use top-level cctools

reckenrode committed:
netlify-cli: use top-level cctools

reckenrode committed:
playwright-test: use top-level cctools

reckenrode committed:
base16-builder: use top-level cctools

reckenrode committed:
mx-puppet-discord: use top-level cctools

reckenrode committed:
mariadb: use top-level cctools

reckenrode committed:
ethercalc: use top-level cctools

reckenrode committed:
make-bootstrap-tools: fix for cctools and ld64 updates

- Copy from darwin.binutils-unwrapped instead of darwin.cctools; and
- Use top-level libtapi.

reckenrode committed:
balena-cli: use top-level cctools

reckenrode committed:
dl-librescore: use top-level cctools

reckenrode committed:
hsd: use top-level cctools

reckenrode committed:
moreutils: use top-level cctools

reckenrode committed:
curl-impersonate: use top-level cctools

reckenrode committed:
onlykey: use top-level cctools

reckenrode committed:
mysql-shell: use top-level cctools

reckenrode committed:
google-amber: use top-level cctools

reckenrode committed:
homepage-dashboard: use top-level cctools

reckenrode committed:
lp_solve: use top-level cctools

reckenrode committed:
mailpit: use top-level cctools

reckenrode committed:
semantic-release: use top-level cctools

reckenrode committed:
unicorn: use top-level cctools

reckenrode committed:
chez: use top-level cctools

reckenrode committed:
defaultGemConfig: use top-level cctools

reckenrode committed:
ruby: use top-level libunwind

reckenrode committed:
valgrind: use top-level cctools

reckenrode committed:
gobject-introspection: use top-level cctools

reckenrode committed:
liquid-dsp: use top-level cctools

reckenrode committed:
mongodb-{5_0,6_0}: use top-level cctools

reckenrode committed:
mysql80: use top-level cctools

reckenrode committed:
rethinkdb: use top-level cctools

reckenrode committed:
shaderc: use top-level cctools

reckenrode committed:
macdylibbundler: use top-level cctools

reckenrode committed:
openvscode-server: use top-level cctools

reckenrode committed:
code-server: use top-level cctools

reckenrode committed:
gargoyle: use top-level cctools

reckenrode committed:
scummvm: use top-level cctools

reckenrode committed:
ghc-standalone-archive: use top-level cctools

reckenrode committed:
ghc: use unwrapped bintools on all platforms

aarch64-darwin no longer needs special handling because `strip` and
`install_name_tool` will update the code signatures of binaries they
modify as long as they are signed by the linker, which is the default.

reckenrode committed:
llvmPackages_{12,13,14,15,16,17,18,git}.libunwind: provide .pc files

Provides compatibility with GNU libunwind for packages that expect to
locate it via pkg-config.

reckenrode committed:
rcodesign: disable impure tests

reckenrode committed:
darwin.AvailabilityVersions: init at 140.1

reckenrode committed:
darwin.Libsystem: get headers from AvailabilityVersions

While these headers correspond to a newer SDK, they’re compatible with the 10.12 SDK and more compatible with GCC.

reckenrode committed:
python3Packages.cffi: fix build with Darwin bootstrap clang

The bootstrap clang reports its version as “boot”, which results in unwanted test failures when the `-Wnull-pointer-subtraction` workaround is disabled.

reckenrode committed:
python3: enable LTO on Darwin

reckenrode committed:
cocotb: enable LTO on Darwin

reckenrode committed:
python3Packages.levenshtein: enable LTO on Darwin

reckenrode committed:
python3Packages.rapidfuzz: enable LTO on Darwin

reckenrode committed:
rubyPackages.sassc: enable LTO on Darwin

reckenrode committed:
tracy: enable LTO on Darwin

reckenrode committed:
rgbds: enable LTO on Darwin

reckenrode committed:
eduke32: enable LTO on Darwin

reckenrode committed:
stockfish: enable LTO on Darwin

reckenrode committed:
firebird-emu: enable LTO on Darwin

reckenrode committed:
timeloop: enable LTO on Darwin

reckenrode committed:
rvvm: enable LTO on Darwin

reckenrode committed:
git: fix Darwin SDK check

The source-based SDK uses a newer version of AvailabilityMacros.h, which
defines various newer SDK availability macros. They are correctly
defined to `UNAVAILABLE_ATTRIBUTE`, but git only checks whether they
exist, which results in its misdetecting the 10.12 SDK as a 10.13 (or
newer) SDK.

reckenrode committed:
SDL2: fix Darwin SDK checks

The source-based SDK uses a newer version of AvailabilityMacros.h, which
defines various newer SDK availability macros. They are correctly
defined to `UNAVAILABLE_ATTRIBUTE`, but git only checks whether they
exist, which results in its misdetecting the 10.12 SDK as a 10.13 (or
newer) SDK.

reckenrode committed:
darwin.apple_sdk: drop NIX_COREFOUNDATION_RPATH

x86_64-darwin no longer uses the source-based CF, making this setup hook
unnecessary and harmful since it can break builds with the updated ld64.

reckenrode committed:
stdenv: set NIX_DONT_SET_RPATH_FOR_TARGET on Darwin

reckenrode committed:
darwin.cctools: delete in favor of top-level cctools

reckenrode committed:
netbsd.compat: use top-level cctools

reckenrode committed:
catboost: use top-level cctools

reckenrode committed:
darwin.stdenv: prepare for Libsystem changes

* Separate Csu and libresolv into separate stage 0 packages to mirror
  how they are organized normally;
* Make bintools overridable for consistency with non-bootstrap bintools;
* Propagate libresolv through the assertions to the final output; and
* Update Libsystem assertions to reflect the fact it will be building
  with `stdenvNoCC` on both x86_64-darwin and aarch64-darwin.

reckenrode committed:
fixDarwinDylibNames: fix targetPrefix when cross-compiling

reckenrode committed:
llvmPackages_16: fix Darwin cross-compilation

Break an infinite recursion with CF due to its depending on libc++ and compiler-rt.

reckenrode committed:
darwin.stdenv: propagate sdkRoot via extraBuildInputs

This ensures it is properly dropped when cross-compiling.

reckenrode committed:
darwin.binutils: refactor and clean up

Update darwin.binutils based on darwin.cctools-llvm. It works
essentially in the same way, simplifying the derivation. The support for
providing both assemblers is handled in a separate commit. Additionally,
unlike cctools-llvm, the lists of tools are provided as passthru
attributes: `cctools_cmds` and `llvm_cmds`.

reckenrode committed:
darwin.{cctools,ld64,libtapi}: remove and add to darwin-aliases

reckenrode commented:
I have pushed an update based on my ld64 branch at https://github.com/NixOS/nixpkgs/pull/307880. It’s a WIP, but it should build. I will be updating this branch regularly along with the ld64 branch. Obviously, this would not go before that branch.

reckenrode committed:
abseil-cpp: only build for the host architecture

abseil-cpp will try to build for both architectures on Darwin. Not sure
why this works with clang 16, but clang 18 fails when x86_64 flags are
used when building for aarch64-darwin.

Et7f3 commented:
I can't build anything with this branch:
<details>
<summary>nix build 'git+file:.?ref=darwin-cross#hello' (infinite recursion encountered)</summary>
```
% nix build 'git+file:.?ref=darwin-cross#hello' --show-trace
error:
       … while calling the 'derivationStrict' builtin

         at /builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'hello-2.12.1'
         whose name attribute is located at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'src' of derivation 'hello-2.12.1'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/by-name/he/hello/package.nix:14:3:

           13|
           14|   src = fetchurl {
             |   ^
           15|     url = "mirror://gnu/hello/hello-${finalAttrs.version}.tar.gz";

       … while evaluating derivation 'hello-2.12.1.tar.gz'
         whose name attribute is located at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'mirrorsFile' of derivation 'hello-2.12.1.tar.gz'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/build-support/fetchurl/default.nix:175:10:

          174|   curlOptsList = lib.escapeShellArgs curlOptsList;
          175|   inherit showURLs mirrorsFile postFetch downloadToTemp executable;
             |          ^
          176|

       … while evaluating derivation 'mirrors-list'
         whose name attribute is located at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'stdenv' of derivation 'mirrors-list'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:360:14:

          359|       args = attrs.args or ["-e" (attrs.builder or ./default-builder.sh)];
          360|       inherit stdenv;
             |              ^
          361|

       … while evaluating derivation 'stdenv-darwin'
         whose name attribute is located at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/default.nix:93:14

       … while evaluating attribute 'allowedRequisites' of derivation 'stdenv-darwin'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/default.nix:84:7:

           83|     lib.optionalAttrs (allowedRequisites != null) {
           84|       allowedRequisites = allowedRequisites
             |       ^
           85|         ++ defaultNativeBuildInputs ++ defaultBuildInputs;

       … while evaluating derivation 'cctools-binutils-darwin-wrapper-1010.6'
         whose name attribute is located at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'bintools_bin' of derivation 'cctools-binutils-darwin-wrapper-1010.6'

         at «none»:0: (source not available)

       … while evaluating derivation 'cctools-binutils-darwin-1010.6'
         whose name attribute is located at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'buildCommand' of derivation 'cctools-binutils-darwin-1010.6'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/os-specific/darwin/binutils/default.nix:63:3:

           62|
           63|   buildCommand = ''
             |   ^
           64|     mkdir -p $out/bin $out/include

       … from call site

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/os-specific/darwin/binutils/default.nix:80:9:

           79|       ln -s "$llvmPath/$llvmTool" "$out/bin/${targetPrefix}$cctoolsTool"
           80|       ${linkManPages llvm-manpages "$llvmTool" "$cctoolsTool"}
             |         ^
           81|     done

       … while calling 'linkManPages'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/os-specific/darwin/binutils/default.nix:42:18:

           41|   linkManPages =
           42|     pkg: source: target:
             |                  ^
           43|     lib.optionalString enableManpages ''

       … from call site

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/os-specific/darwin/binutils/default.nix:43:5:

           42|     pkg: source: target:
           43|     lib.optionalString enableManpages ''
             |     ^
           44|       sourcePath=${pkg}/share/man/man1/${source}.1.gz

       … while calling 'optionalString'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/lib/strings.nix:268:5:

          267|     # String to return if condition is true
          268|     string: if cond then string else "";
             |     ^
          269|

       … while evaluating derivation 'llvm-manpages-16.0.6'
         whose name attribute is located at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute '__impureHostDeps' of derivation 'llvm-manpages-16.0.6'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:446:7:

          445|       __propagatedSandboxProfile = unique (computedPropagatedSandboxProfile ++ [ propagatedSandboxProfile ]);
          446|       __impureHostDeps = computedImpureHostDeps ++ computedPropagatedImpureHostDeps ++ __propagatedImpureHostDeps ++ __impureHostDeps ++ stdenv.__extraImpureHostDeps ++ [
             |       ^
          447|         "/dev/zero"

       … while calling anonymous lambda

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:430:30:

          429|         computedImpureHostDeps =
          430|           unique (concatMap (input: input.__propagatedImpureHostDeps or [])
             |                              ^
          431|             (stdenv.extraNativeBuildInputs

       … while calling anonymous lambda

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:436:30:

          435|         computedPropagatedImpureHostDeps =
          436|           unique (concatMap (input: input.__propagatedImpureHostDeps or [])
             |                              ^
          437|             allPropagatedDependencies);

       … while calling anonymous lambda

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:321:13:

          320|       (map (drv: getDev drv.__spliced.hostHost or drv) (checkDependencyList "depsHostHostPropagated" depsHostHostPropagated))
          321|       (map (drv: getDev drv.__spliced.hostTarget or drv) (checkDependencyList "propagatedBuildInputs" propagatedBuildInputs))
             |             ^
          322|     ]

       … from call site

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:321:18:

          320|       (map (drv: getDev drv.__spliced.hostHost or drv) (checkDependencyList "depsHostHostPropagated" depsHostHostPropagated))
          321|       (map (drv: getDev drv.__spliced.hostTarget or drv) (checkDependencyList "propagatedBuildInputs" propagatedBuildInputs))
             |                  ^
          322|     ]

       … while calling 'getOutput'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/lib/attrsets.nix:1796:23:

         1795|   */
         1796|   getOutput = output: pkg:
             |                       ^
         1797|     if ! pkg ? outputSpecified || ! pkg.outputSpecified

       … while calling anonymous lambda

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/lib/lists.nix:334:29:

          333|   */
          334|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                             ^
          335|

       … from call site

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/lib/lists.nix:334:32:

          333|   */
          334|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                                ^
          335|

       … while calling anonymous lambda

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:276:15:

          275|     imap1
          276|       (index: dep:
             |               ^
          277|         if isDerivation dep || dep == null || builtins.isString dep || builtins.isPath dep then dep

       … from call site

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:277:12:

          276|       (index: dep:
          277|         if isDerivation dep || dep == null || builtins.isString dep || builtins.isPath dep then dep
             |            ^
          278|         else if isList dep then checkDependencyList' ([index] ++ positions) name dep

       … while calling 'isDerivation'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/lib/attrsets.nix:1283:5:

         1282|   isDerivation =
         1283|     value: value.type or null == "derivation";
             |     ^
         1284|

       … while calling 'checkDrv'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/development/interpreters/python/mk-python-derivation.nix:206:26:

          205|
          206|     checkDrv = attrName: drv:
             |                          ^
          207|       if (isPythonModule drv) && (isMismatchedPython drv) then throwMismatch attrName drv

       … from call site

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/development/interpreters/python/mk-python-derivation.nix:207:35:

          206|     checkDrv = attrName: drv:
          207|       if (isPythonModule drv) && (isMismatchedPython drv) then throwMismatch attrName drv
             |                                   ^
          208|       else drv;

       … while calling 'isMismatchedPython'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/development/interpreters/python/mk-python-derivation.nix:48:24:

           47|
           48|   isMismatchedPython = drv: drv.pythonModule != python;
             |                        ^
           49|

       … while evaluating derivation 'python3-3.11.9'
         whose name attribute is located at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'CPPFLAGS' of derivation 'python3-3.11.9'

         at «none»:0: (source not available)

       … while calling anonymous lambda

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:576:11:

          575|     mapAttrs
          576|       (n: v: assert assertMsg (isString v || isBool v || isInt v || isDerivation v)
             |           ^
          577|         "The ‘env’ attribute set can only contain derivation, string, boolean or integer attributes. The ‘${n}’ attribute is of type ${builtins.typeOf v}."; v)

       … from call site

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:576:21:

          575|     mapAttrs
          576|       (n: v: assert assertMsg (isString v || isBool v || isInt v || isDerivation v)
             |                     ^
          577|         "The ‘env’ attribute set can only contain derivation, string, boolean or integer attributes. The ‘${n}’ attribute is of type ${builtins.typeOf v}."; v)

       … while calling 'assertMsg'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/lib/asserts.nix:41:5:

           40|     pred:
           41|     msg:
             |     ^
           42|     pred || builtins.throw msg;

       … while calling anonymous lambda

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/development/interpreters/python/cpython/default.nix:375:43:

          374|   env = {
          375|     CPPFLAGS = concatStringsSep " " (map (p: "-I${getDev p}/include") buildInputs);
             |                                           ^
          376|     LDFLAGS = concatStringsSep " " (map (p: "-L${getLib p}/lib") buildInputs);

       … while evaluating derivation 'configd-453.19'
         whose name attribute is located at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'stdenv' of derivation 'configd-453.19'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:360:14:

          359|       args = attrs.args or ["-e" (attrs.builder or ./default-builder.sh)];
          360|       inherit stdenv;
             |              ^
          361|

       … while evaluating derivation 'bootstrap-stage4-stdenv-darwin'
         whose name attribute is located at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/default.nix:93:14

       … while evaluating attribute 'defaultNativeBuildInputs' of derivation 'bootstrap-stage4-stdenv-darwin'

         at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/default.nix:121:14:

          120|
          121|       inherit initialPath shell
             |              ^
          122|         defaultNativeBuildInputs defaultBuildInputs;

       … while evaluating derivation 'clang-wrapper-16.0.6'
         whose name attribute is located at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'bintools' of derivation 'clang-wrapper-16.0.6'

         at «none»:0: (source not available)

       … while evaluating derivation 'cctools-binutils-darwin-wrapper-1010.6'
         whose name attribute is located at /nix/store/sz8cwh05d5v895hysgishb7k094r6d7p-source/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'bintools_bin' of derivation 'cctools-binutils-darwin-wrapper-1010.6'

         at «none»:0: (source not available)

       error: infinite recursion encountered

       at «none»:0: (source not available)
```
</details>

reckenrode commented:
I’ll try to push an update this weekend. There’s an issue with man pages that should be fixed on my working branch. I’ve been holding off pushing an update because I’d like to push updates only when the Darwin channel blockers build, but the branch now includes work preparing for the clang 18 update, which is surprisingly disruptive (e.g., it breaks building GCC as well as older versions of clang, and I haven’t finished fixing that out yet).

reckenrode committed:
curl: enable static builds on x86_64-darwin

reckenrode committed:
xorg-xserver: fix build with LLVM libunwind

reckenrode committed:
meson: use _accumFlagsArray

reckenrode committed:
darwin.Csu: don’t override startup object deployment targets

GCC has been updated not to pass `-rpaths` because the updated ld64 won’t accept them for older deployment targets regardless of what the target is of the object files, so this workaround is no longer needed.

reckenrode committed:
libpsl: fix cross-compilation

libpsl needs to link against gettext, which requires it be included as a build input even if gettext is already propagated as a native build input.

reckenrode committed:
http-parser: fix cross-compilation

The http-parser `Makefile` excepts the variable to be `PLATFORM`. Otherwise, it runs `uname -s` and detects the wrong target platform when cross-compiling to Linux from Darwin.

reckenrode committed:
boost: fix Darwin cross-compilation

b2 tries to invoke the linker without a prefix, which fails.

reckenrode committed:
boost-build: fix aarch64 clang cross-compilation

reckenrode committed:
python3: fix static builds on Darwin

Actual static builds using `-static` are not supported, but Darwin can still do mostly static.

reckenrode commented:
I’ll have an updated branch sometime this week. I created a prerequisite and related PRs branch for tracking fixes. I was hoping to have the branch tomorrow, but then I found the libiconv issue that is going to cause me to rebuild almost everything again for my tests.

reckenrode commented:
Latest force push is my branch up until right before the clang 18 stuff.

reckenrode commented:
Sorry about not updating, but this PR will be replaced by the Darwin refactor. I have a replacement for the SDKs that makes using them much easier in Darwin. I am developing it with cross-compilation in mind, so it should just work. This work is targeted for 24.11.

Mindavi commented:
Thanks for your hard work on that. I think we can close this then, for now.

will commented:
> Sorry about not updating, but this PR will be replaced by the Darwin refactor. I have a replacement for the SDKs that makes using them much easier in Darwin. I am developing it with cross-compilation in mind, so it should just work. This work is targeted for 24.11.

Is there a link to somewhere (pr, or message board thread, or anything) to follow along with this? Just curious to see what's going on

reckenrode commented:
> Is there a link to somewhere (pr, or message board thread, or anything) to follow along with this? Just curious to see what's going on

Discussions have been happening on Matrix and in other PRs. I discuss it some in https://github.com/NixOS/nixpkgs/issues/242666#issuecomment-2282854347. To help explain further and provide some background, I created a thread on Discourse: https://discourse.nixos.org/t/on-the-future-of-darwin-sdks-or-how-you-can-stop-worrying-and-put-the-sdk-in-build-inputs/50574
